shared/js/intl_helper.js

```diff
@@ -0,0 +1,170 @@
+        }
+
+        return 'hour' in obj.options;
+      }
+    },
+    'number': {
+      set: function(options) {
+        return Intl.NumberFormat(
+          navigator.languages,
+          options
+        );
+      },
+      isAffected: function(obj, reason) {
+        return true;
+      }
+    },
+    'collator': {
+      set: function(options) {
+        return Intl.Collator(
+          navigator.languages,
+          options
+        );
+      },
+      isAffected: function(obj, reason) {
+        return true;
+      }
+    },
+  };
+
+  const affectedTypes = {
+    'languagechange': ['datetime', 'number', 'collator'],
+    'timeformatchange': ['datetime'],
+  };
+
+  global.IntlHelper = {
+    define: function(name, type, options) {
+      if (!knownObjects.hasOwnProperty(type)) {
+        throw new Error('Unknown type: ' + type);
+      }
+
+      if (!objectCache) {
+        objectCache = new Map();
+      }
+
+      if (objectCache.has(name)) {
+        throw new Error(
+          'Intl Object with name "' + name + '" already exists');
+      }
+
+      objectCache.set(name, {
+        options,
+        type,
+        object: undefined,
+        listeners: new Set()
+      });
+    },
+
+    get: function(name) {
+      if (!objectCache.has(name)) {
+        throw new Error('Intl Object with name "' + name + '" is not defined');
+      }
+      const obj = objectCache.get(name);
+      
+      if (!knownObjects.hasOwnProperty(obj.type)) {
+        throw new Error('Unknown type: ' + obj.type);
+      }
+
+      if (obj.object === undefined) {
+        obj.object = knownObjects[obj.type].set(obj.options);
+      }
+      return obj.object;
+    },
+
+    observe: function(names, cb) {
+      if (typeof names === 'string') {
+        names = [names];
+      }
+      for (var name of names) {
+        if (!objectCache.has(name)) {
+          throw new Error(
+            'Intl Object with name "' + name + '" is not defined');
+        }
+
+        var obj = objectCache.get(name);
+
+        obj.listeners.add(cb);
+      }
+    },
+
+    unobserve: function(name, cb) {
+      var obj = objectCache.get(name);
+
+      obj.listeners.delete(cb);
+    },
+
+    _fireObservers: function(affectedObjects, ...args) {
+      var affectedCallbacks = new Set();
+
+      affectedObjects.forEach((obj, name) => {
+        obj.listeners.forEach(cb => affectedCallbacks.add(cb));
+      });
+
+      affectedCallbacks.forEach(cb => cb(...args));
+    },
+
+    handleEvent: function(evt) {
+      const affectedObjects = resetObjects(affectedTypes[evt.type], evt.type);
+
+      global.IntlHelper._fireObservers(affectedObjects);
+    },
+
+    _resetObjectCache: function() {
+      objectCache = undefined;
+    }
+  };
+
+  function resetObjects(types, reason) {
+    const affectedObjects = new Set();
+
+    objectCache.forEach((obj, name) => {
+      if (types.indexOf(obj.type) !== -1 &&
+          knownObjects[obj.type].isAffected(obj, reason)) {
```

Comment:
Yeah, since you're iterating over `objectCache` anyways here, I'd just remove `affectedTypes` altogether and rely on `isAffected` to do the filtering.